diff --git a/color-rg.el b/color-rg.el
index c186c92..6bd965e 100644
--- a/color-rg.el
+++ b/color-rg.el
@@ -621,6 +621,11 @@ This function is called from `compilation-filter-hook'."
                     (run-at-time "1sec" nil
                                  (lambda ()
                                    (color-rg-clone-to-temp-buffer)))
+
+                    (read-only-mode -1)
+                    (ansi-color-apply-on-region (point-min) (point-max))
+                    (read-only-mode 1)
+
                     `(,(format "finished (%d matches found)\n" color-rg-hit-count) . "matched"))
                    ((not (buffer-modified-p))
                     '("finished with no matches found\n" . "no match"))
@@ -877,12 +882,14 @@ user more freedom to use rg with special arguments."
 (defun color-rg-read-input ()
   (let* ((current-symbol (color-rg-pointer-string))
          (input-string
-          (string-trim
-           (read-string
-            (format "COLOR-RG Search (%s): " current-symbol)
-            nil
-            'color-rg-read-input-history
-            ))))
+          (if (use-region-p)
+              ""
+            (string-trim
+             (read-string
+              (format "COLOR-RG Search (%s): " current-symbol)
+              nil
+              'color-rg-read-input-history
+              )))))
     (when (string-blank-p input-string)
       (setq input-string current-symbol))
     input-string))
@@ -1198,17 +1205,19 @@ This assumes that `color-rg-in-string-p' has already returned true, i.e.
   ;; Set `enable-local-variables' to :safe, avoid emacs ask annoyingly question when open file by color-rg.
   (setq enable-local-variables :safe)
   ;; Search.
-  (let ((search-keyword
-         (or keyword
-             (color-rg-read-input)))
-        (search-directory
-         (expand-file-name
-          (or directory
-              default-directory)))
-        (search-globs
-         (or globs
-             "everything")))
-    (color-rg-search search-keyword
+  (let* ((search-keyboard
+          (or keyword
+              (color-rg-read-input)))
+         (search-directory
+          (expand-file-name
+           (or directory
+               (if (featurep 'projectile)
+                   (projectile-project-root))
+               default-directory)))
+         (search-globs
+          (or globs
+              "everything")))
+    (color-rg-search search-keyboard
                      search-directory
                      search-globs)))
 
