diff --git a/color-rg.el b/color-rg.el
index 6894362..e23e80e 100644
--- a/color-rg.el
+++ b/color-rg.el
@@ -621,6 +621,11 @@ This function is called from `compilation-filter-hook'."
                     (run-at-time "1sec" nil
                                  (lambda ()
                                    (color-rg-clone-to-temp-buffer)))
+
+                    (read-only-mode -1)
+                    (ansi-color-apply-on-region (point-min) (point-max))
+                    (read-only-mode 1)
+
                     `(,(format "finished (%d matches found)\n" color-rg-hit-count) . "matched"))
                    ((not (buffer-modified-p))
                     '("finished with no matches found\n" . "no match"))
@@ -871,12 +876,14 @@ user more freedom to use rg with special arguments."
 (defun color-rg-read-input ()
   (let* ((current-symbol (color-rg-pointer-string))
          (input-string
-          (string-trim
-           (read-string
-            (format "COLOR-RG Search (%s): " current-symbol)
-            nil
-            'color-rg-read-input-history
-            ))))
+          (if (use-region-p)
+              ""
+            (string-trim
+             (read-string
+              (format "COLOR-RG Search (%s): " current-symbol)
+              nil
+              'color-rg-read-input-history
+              )))))
     (when (string-blank-p input-string)
       (setq input-string current-symbol))
     input-string))
@@ -1197,6 +1204,8 @@ This assumes that `color-rg-in-string-p' has already returned true, i.e.
               (color-rg-read-input)))
          (search-directory
           (or directory
+              (if (featurep 'projectile)
+                  (projectile-project-root))
               default-directory))
          (search-globs
           (or globs
@@ -1227,8 +1236,8 @@ This assumes that `color-rg-in-string-p' has already returned true, i.e.
     (if project
         (expand-file-name
          (cond
-          ((fboundp 'project-root) (project-root project))
-          ((fboundp 'project-roots) (car (project-roots project)))))
+          ((fboundp project-root) (project-root project))
+          ((fboundp project-roots) (car (project-roots project)))))
       default-directory)))
 
 (defalias 'color-rg-search-input-in-project 'color-rg-search-project)
