diff --git a/copilot.el b/copilot.el
index 0fba341..057c862 100644
--- a/copilot.el
+++ b/copilot.el
@@ -253,9 +253,26 @@ Username and password are optional."
    (t
     (concat "file://" (url-encode-url buffer-file-name)))))
 
+(defvar copilot-max-char 30000)
+(defvar-local copilot-line-bias 1)
+
 (defun copilot--generate-doc ()
   "Generate doc parameters for completion request."
-  (list :source (concat (buffer-substring-no-properties (point-min) (point-max)) "\n")
+  (setq-local copilot-line-bias 1)
+  (list :source (concat
+                 (if (> (point-max) copilot-max-char)
+                     (cond
+                      ((< (- (point-max) (point)) (/ copilot-max-char 2))
+                       (setq-local copilot-line-bias (line-number-at-pos (- (point-max) copilot-max-char)))
+                       (buffer-substring-no-properties (- (point-max) copilot-max-char) (point-max)))
+                      ((< (- (point) (point-min)) (/ copilot-max-char 2))
+                       (buffer-substring-no-properties (point-min) (+ (point-min) copilot-max-char)))
+                      (t
+                       (setq-local copilot-line-bias (line-number-at-pos (- (point) (/ copilot-max-char 2))))
+                       (buffer-substring-no-properties (- (point) (/ copilot-max-char 2))
+                                                       (+ (point) (/ copilot-max-char 2)))))
+                   (buffer-substring-no-properties (point-min) (point-max)))
+                  "\n")
         :tabSize (copilot--infer-indentation-offset)
         :indentSize (copilot--infer-indentation-offset)
         :insertSpaces (if indent-tabs-mode :json-false t)
@@ -263,9 +280,10 @@ Username and password are optional."
         :uri (copilot--get-uri)
         :relativePath (copilot--get-relative-path)
         :languageId (s-chop-suffix "-mode" (symbol-name major-mode))
-        :position (list :line (1- (line-number-at-pos))
+        :position (list :line (- (line-number-at-pos) copilot-line-bias)
                         :character (- (point) (point-at-bol)))))
 
+
 (defun copilot--get-completion (callback)
   "Get completion with CALLBACK."
   (copilot--async-request 'getCompletions
@@ -341,6 +359,7 @@ To work around posn problems with after-string property.")
 For Copilot, COL is always 0.
 USER-POS is the cursor position (for verification only)."
   (copilot-clear-overlay)
+  (setq line (1- (+ line copilot-line-bias)))
   (save-excursion
     (widen)
     (goto-char (point-min))
