diff --git a/acm/acm-backend-codeium.el b/acm/acm-backend-codeium.el
index 32dcc10..f87cce2 100644
--- a/acm/acm-backend-codeium.el
+++ b/acm/acm-backend-codeium.el
@@ -160,6 +160,7 @@
 (defun acm-backend-codeium-candidate-expand (candidate-info bound-start &optional preview)
   ;; We need replace whole area with codeium label.
   (let ((end-position (line-end-position)))
+    (setq end-position (point))
     (forward-line (- (plist-get candidate-info :line) (count-lines (point-min) (line-beginning-position))))
     (if preview
         (acm-preview-create-overlay (point) end-position (plist-get candidate-info :label))
diff --git a/acm/acm-backend-copilot.el b/acm/acm-backend-copilot.el
index f038ae1..048b860 100644
--- a/acm/acm-backend-copilot.el
+++ b/acm/acm-backend-copilot.el
@@ -61,6 +61,7 @@ in the proxy plist. For example:
 (defun acm-backend-copilot-candidate-expand (candidate-info bound-start &optional preview)
   ;; We need replace whole area with copilot label.
   (let ((end-position (line-end-position)))
+    (setq end-position (point))
     (forward-line (- (plist-get candidate-info :line) (count-lines (point-min) (line-beginning-position))))
     (if preview
         (acm-preview-create-overlay (point) end-position (plist-get candidate-info :label))
diff --git a/acm/acm-frame.el b/acm/acm-frame.el
index abc9846..8be0225 100644
--- a/acm/acm-frame.el
+++ b/acm/acm-frame.el
@@ -155,13 +155,15 @@
                 ))
   (add-function :after after-focus-change-function #'acm-frame-restore-font))
 
-(cl-defmacro acm-frame-new (frame frame-buffer buffer-name &optional max-width max-height popup-pos)
+(cl-defmacro acm-frame-new (frame frame-buffer buffer-name &optional max-width max-height popup-pos margin)
   `(progn
      (when (and (frame-live-p ,frame)
                 (not (eq (frame-parent ,frame) (selected-frame))))
        (acm-frame-delete-frame ,frame))
 
-     (acm-frame-create-frame-if-not-exist ,frame ,frame-buffer ,buffer-name 1 t)
+     (if ,margin
+         (acm-frame-create-frame-if-not-exist ,frame ,frame-buffer ,buffer-name ,margin t)
+       (acm-frame-create-frame-if-not-exist ,frame ,frame-buffer ,buffer-name 1 t))
 
      (acm-frame-set-frame-max-size ,frame ,max-width ,max-height)
 
diff --git a/acm/acm.el b/acm/acm.el
index 7c1f9d6..462ef06 100644
--- a/acm/acm.el
+++ b/acm/acm.el
@@ -219,7 +219,8 @@
     (define-key map "\M-." #'acm-select-first)
     (define-key map "\C-v" #'acm-select-next-page)
     (define-key map "\M-v" #'acm-select-prev-page)
-    (define-key map [tab]  #'acm-complete)
+    (define-key map [tab]  #'acm-select-next)
+    (define-key map [backtab]  #'acm-select-prev)
     (define-key map "\C-m" #'acm-complete)
     (define-key map "\t" #'acm-complete)
     (define-key map "\n" #'acm-complete)
@@ -231,8 +232,8 @@
     (define-key map "\M-k" #'acm-doc-scroll-down)
     (define-key map "\M-l" #'acm-hide)
     (define-key map "\C-g" #'acm-hide)
-    (define-key map "1" #'acm-insert-number-or-complete-candiate)
-    (define-key map "2" #'acm-insert-number-or-complete-candiate)
+    ;;(define-key map "1" #'acm-insert-number-or-complete-candiate)
+    ;;(define-key map "2" #'acm-insert-number-or-complete-candiate)
     (define-key map "3" #'acm-insert-number-or-complete-candiate)
     (define-key map "4" #'acm-insert-number-or-complete-candiate)
     (define-key map "5" #'acm-insert-number-or-complete-candiate)
@@ -240,7 +241,7 @@
     (define-key map "7" #'acm-insert-number-or-complete-candiate)
     (define-key map "8" #'acm-insert-number-or-complete-candiate)
     (define-key map "9" #'acm-insert-number-or-complete-candiate)
-    (define-key map "0" #'acm-insert-number-or-complete-candiate)
+    ;;(define-key map "0" #'acm-insert-number-or-complete-candiate)
     map)
   "Keymap used when popup is shown.")
 
@@ -523,6 +524,12 @@ Only calculate template candidate when type last character."
               (max (- acm-menu-length (+ (length template-first-part-candidates) (length tabnine-candidates)))
                    mode-candidates-min-index))
 
+        ;; always show codeium candidates
+        (when codeium-candidates
+          (setq mode-candidates-split-index
+                (max (- mode-candidates-split-index (length codeium-candidates))
+                     mode-candidates-min-index)))
+
         ;; Build mode candidates.
         (if (> (length mode-candidates) mode-candidates-split-index)
             (progn
@@ -874,7 +881,7 @@ The key of candidate will change between two LSP results."
                         (+ cursor-y offset-y)))))
     (acm-frame-set-frame-position acm-menu-frame acm-frame-x acm-frame-y)))
 
-(defun acm-doc-try-show (&optional update-completion-item)
+(defun acm-doc-try-show2 (&optional update-completion-item)
   (when acm-enable-doc
     (let* ((candidate (acm-menu-current-candidate))
            (backend (plist-get candidate :backend))
@@ -918,6 +925,19 @@ The key of candidate will change between two LSP results."
           ;; Hide doc frame immediately if backend is not LSP.
           (_ (acm-doc-hide)))))))
 
+(defvar acm-doc-show-timer nil)
+(defvar acm-doc-show-delay 0.5)
+
+(defun acm-doc-try-show (&optional update-completion-item)
+  "Show doc frame."
+  (acm-doc-hide)
+  (acm-cancel-timer acm-doc-show-timer)
+  (setq acm-doc-show-timer
+        (run-with-idle-timer acm-doc-show-delay nil
+                             (lambda ()
+                               (if (frame-visible-p acm-menu-frame)
+                                   (acm-doc-try-show2 update-completion-item))))))
+
 (defun acm-doc-frame-adjust ()
   (let* ((emacs-width (frame-pixel-width))
          (emacs-height (frame-pixel-height))
diff --git a/core/codeium.py b/core/codeium.py
index a680456..23f5ad2 100644
--- a/core/codeium.py
+++ b/core/codeium.py
@@ -23,6 +23,8 @@ import time
 import traceback
 import urllib.parse
 import urllib.request
+from distutils.version import StrictVersion
+import threading
 
 from core.utils import *
 
@@ -44,29 +46,22 @@ class Codeium:
         self.counter = 1
         self.wait_request = []
 
-    def complete(
-        self,
-        cursor_offset,
-        editor_language,
-        tab_size,
-        text,
-        insert_spaces,
-        prefix,
-        language,
-    ):
+    def cancel_copmlete(self):
+        for _ in self.wait_request:
+            self.metadata['request_id'] = self.wait_request.pop()
+            self.post_request(self.make_url('CancelRequest'), {'metadata': self.metadata})
+
+    def complete(self, cursor_offset, editor_language, tab_size, text, insert_spaces, prefix, language):
         self.get_info()
         self.run_local_server()
 
+        thread = threading.Thread(target=self.cancel_copmlete)
+        thread.start()
+
         # utf-8 cursor offset
         cursor_offset = len(text[:cursor_offset].encode("utf-8", errors="ignore"))
         self.current_cussor_offset = cursor_offset
 
-        for _ in self.wait_request:
-            self.metadata["request_id"] = self.wait_request.pop()
-            self.post_request(
-                self.make_url("CancelRequest"), {"metadata": self.metadata}
-            )
-
         self.metadata["request_id"] = self.counter
         self.wait_request.append(self.counter)
         self.counter += 1
@@ -144,16 +139,12 @@ class Codeium:
                 labels = label.strip().split("\n")
                 first_line = labels[0]
 
-                document = f"```{language}\n{label}\n```"
-
-                # Don't make display label bigger than max length.
-                display_label = first_line
-                if len(first_line) > self.display_label_max_length:
-                    display_label = "... " + display_label[len(first_line) - self.display_label_max_length:]
+                # document = f"```{language}\n{label}\n```" if len(labels) > 1 else ""
+                document = label if len(labels) > 1 else ""
 
-                # Only hide documentation when label smaller than max length and only 1 line
-                if len(labels) <= 1 and len(first_line) <= self.display_label_max_length:
-                        document = ""
+                display_label = labels[0]
+                if len(display_label) > self.display_label_max_length or len(labels) > 1:
+                        display_label = display_label[:self.display_label_max_length - 4] + " ..."
 
                 completion_parts = completion.get("completionParts", [{}])[0]
                 annotation = (
diff --git a/core/fileaction.py b/core/fileaction.py
index 533d3b1..8de6267 100755
--- a/core/fileaction.py
+++ b/core/fileaction.py
@@ -203,15 +203,19 @@ class FileAction:
         self.last_change_file_time = time.time()
 
         # Send textDocument/completion 100ms later.
-        delay = 0 if is_running_in_server() else 0.1
-        self.try_completion_timer = threading.Timer(delay, lambda : self.try_completion(position, before_char, prefix, self.version))
-        self.try_completion_timer.start()
+        if position['line'] < 0:
+            print("error in change file", start, end, range_length, change_text, position, before_char, buffer_name, prefix, self.org_line_bias)
+        else:
+            delay = 0 if is_running_in_server() else 0.1
+            self.try_completion_timer = threading.Timer(delay, lambda : self.try_completion(position, before_char, prefix, self.version))
+            self.try_completion_timer.start()
 
     def update_file(self, buffer_name, org_line_bias=None):
         self.org_line_bias = org_line_bias
         buffer_content = get_buffer_content(self.filepath, buffer_name)
         for lsp_server in self.get_lsp_servers():
             lsp_server.send_whole_change_notification(self.filepath, self.version, buffer_content)
+            print(lsp_server, 'update_file')
         self.version += 1
 
     def try_completion(self, position, before_char, prefix, version=None):
@@ -410,7 +414,13 @@ class FileAction:
     def rename_file(self, old_filepath, new_filepath):
         self.get_lsp_servers()[0].send_did_rename_files_notification(old_filepath, new_filepath)
 
+    completion_request_ids = -1
     def send_server_request(self, lsp_server, handler_name, *args, **kwargs):
+        if self.completion_request_ids > 0 and handler_name == "completion":
+            # Drop completion request if previous completion request is not finished.
+            print("Drop completion request for {}".format(self.filepath), self.completion_request_ids)
+            # return
+
         handler: Handler = self.method_handlers[lsp_server.server_info["name"]][handler_name]
 
         handler.latest_request_id = request_id = generate_request_id()
@@ -427,6 +437,10 @@ class FileAction:
             params=params,
             request_id=request_id)
 
+        if handler_name == "completion":
+            self.completion_request_ids = request_id
+            self.completion_start_time = time.time()
+        
     def exit(self):
         for lsp_server in (self.org_lang_servers.values() if self.org_file else self.get_lsp_servers()):
             if lsp_server.server_name in LSP_SERVER_DICT:
diff --git a/core/lspserver.py b/core/lspserver.py
index 3afab18..f29dec8 100755
--- a/core/lspserver.py
+++ b/core/lspserver.py
@@ -654,6 +654,14 @@ class LspServer:
             else:
                 if "method" not in message and message["id"] in self.request_dict:
                     handler = self.request_dict[message["id"]]
+                    if handler.method == 'textDocument/completion':
+                        for key  in self.files:
+                            fileaction = self.files[key]
+                            if message['id'] == fileaction.completion_request_ids:
+                                fileaction.completion_request_ids = -1
+                                import time
+                                print(fileaction, message['id'],
+                                      time.time() - fileaction.completion_start_time)
                     handler.handle_response(
                         request_id=message["id"],
                         response=message["result"],
diff --git a/core/search_file_words.py b/core/search_file_words.py
index b23884b..095475a 100755
--- a/core/search_file_words.py
+++ b/core/search_file_words.py
@@ -80,7 +80,12 @@ class SearchFileWords:
             return
 
         if buffer_name in self.search_content_dict:
-            content = rebuild_content_from_diff(self.search_content_dict[buffer_name], start_pos, end_pos, change_text)
+            try:
+                content = rebuild_content_from_diff(self.search_content_dict[buffer_name], start_pos, end_pos, change_text)
+            except:
+                print('change_buffer error: buffer_name: {}, start_pos: {}, end_pos: {}, change_text: {}'.format(buffer_name, start_pos, end_pos, change_text))
+                return
+                #assert 0
         else:
             content = get_emacs_func_result('get-buffer-content', buffer_name, True)
 
@@ -145,12 +150,16 @@ class SearchFileWords:
                 else:
                     break
         else:
-            match_words = list(filter(lambda word: word.lower().startswith(prefix.lower()), all_words))
-            if prefix.isupper():
+            sprefix = prefix.split('.')[-1]
+            match_words = list(filter(lambda word: word.lower().startswith(sprefix.lower()), all_words))
+            candidates = []
+            if prefix.isupper() and '.' not in prefix:
                 candidates = list(map(lambda word: word.upper(), match_words))
             else:
-                candidates = list(map(lambda word: prefix + word[len(prefix):], match_words))
+                candidates = list(map(lambda word: prefix + word[len(sprefix):], match_words))
+
             candidates.sort(key=len, reverse=False)
+
         return candidates
             
     def search_dispatcher(self):
diff --git a/langserver/pyright-background-analysis.json b/langserver/pyright-background-analysis.json
index 06d7c3c..92fb9e2 100644
--- a/langserver/pyright-background-analysis.json
+++ b/langserver/pyright-background-analysis.json
@@ -11,7 +11,7 @@
       "autoImportCompletions": true,
       "typeshedPaths": [],
       "stubPath": "",
-      "useLibraryCodeForTypes": true,
+      "useLibraryCodeForTypes": false,
       "diagnosticMode": "openFilesOnly",
       "typeCheckingMode": "basic",
       "logLevel": "verbose",
diff --git a/lsp-bridge-diagnostic.el b/lsp-bridge-diagnostic.el
index 5394a67..4d7b694 100644
--- a/lsp-bridge-diagnostic.el
+++ b/lsp-bridge-diagnostic.el
@@ -170,7 +170,16 @@ You can set this value with `(2 3 4) if you just need render error diagnostic."
 
 (defvar lsp-bridge-diagnostic-display-error-at-point-timer nil)
 
-(autoload 'lsp-bridge--with-file-buffer "lsp-bridge")
+;;(autoload 'lsp-bridge--with-file-buffer "lsp-bridge")
+
+(cl-defmacro lsp-bridge--with-file-buffer (filename filehost &rest body)
+  "Evaluate BODY in buffer with FILEPATH."
+  (declare (indent 1))
+  `(when-let ((buffer (pcase filehost
+                        ("" (lsp-bridge-get-match-buffer-by-filepath ,filename))
+                        (_ (lsp-bridge-get-match-buffer-by-remote-file ,filehost ,filename)))))
+     (with-current-buffer buffer
+       ,@body)))
 
 (defun lsp-bridge-diagnostic-hide-tooltip ()
   (acm-frame-hide-frame lsp-bridge-diagnostic-frame))
@@ -263,15 +272,17 @@ You can set this value with `(2 3 4) if you just need render error diagnostic."
     ;; Show diagnostic tooltip.
     (with-current-buffer (get-buffer-create lsp-bridge-diagnostic-buffer)
       (erase-buffer)
-      (lsp-bridge-diagnostic-insert-colored-string (overlay-get diagnostic-overlay 'color) diagnostic-display-message)
+      ;;(lsp-bridge-diagnostic-insert-colored-string (overlay-get diagnostic-overlay 'color) diagnostic-display-message)
+      (insert (propertize diagnostic-display-message 'face 'flycheck-fringe-warning))
       (setq-local lsp-bridge-diagnostic-message diagnostic-message))
 
     (cond
      ((acm-frame-can-display-p)
       (acm-frame-new lsp-bridge-diagnostic-frame
                      lsp-bridge-diagnostic-buffer
-                     "lsp bridge diagnostic frame")
-
+                     "lsp bridge diagnostic frame"
+                     ;; zero margin
+                     nil nil nil 0)
       (set-face-background (acm-frame-border-face)
                            foreground-color
                            lsp-bridge-diagnostic-frame))
diff --git a/lsp-bridge.el b/lsp-bridge.el
index ee3ba19..4b3b92d 100644
--- a/lsp-bridge.el
+++ b/lsp-bridge.el
@@ -893,23 +893,27 @@ So we build this macro to restore postion after code format."
             (lsp-bridge-enable-org-babel
              (lsp-bridge-org-babel-check-lsp-server)))))))
 
+(defvar-local lsp-bridge-has-lsp-server--p nil)
+(defvar-local lsp-bridge-has-lsp-server--p-set nil)
 (defun lsp-bridge-has-lsp-server-p ()
   (cond ((and lsp-bridge-enable-org-babel (eq major-mode 'org-mode))
          (setq-local acm-is-elisp-mode-in-org nil)
          (lsp-bridge-org-babel-check-lsp-server))
+        (lsp-bridge-has-lsp-server--p-set lsp-bridge-has-lsp-server--p)
         (t
-         (when-let* ((filename (or (ignore-errors (file-truename
-                                                   (lsp-bridge-get-buffer-file-name-text)))
-                                   (when (lsp-bridge-is-remote-file)
-                                     lsp-bridge-remote-file-path))))
-           (let* ((multi-lang-server-by-extension (or (lsp-bridge-get-multi-lang-server-by-extension filename)
-                                                      (lsp-bridge-get-multi-lang-server-by-file-mode filename)))
-                  (lang-server-by-extension (or (lsp-bridge-get-single-lang-server-by-extension filename)
-                                                (lsp-bridge-get-single-lang-server-by-file-mode filename))))
-             (if multi-lang-server-by-extension
-                 multi-lang-server-by-extension
-               lang-server-by-extension)
-             )))))
+         (setq-local lsp-bridge-has-lsp-server--p-set t)
+         (setq-local lsp-bridge-has-lsp-server--p
+                     (when-let* ((filename (or (ignore-errors (file-truename buffer-file-name))
+                                               (when (lsp-bridge-is-remote-file)
+                                                 lsp-bridge-remote-file-path))))
+                       (let* ((multi-lang-server-by-extension (or (lsp-bridge-get-multi-lang-server-by-extension filename)
+                                                                  (lsp-bridge-get-multi-lang-server-by-file-mode filename)))
+                              (lang-server-by-extension (or (lsp-bridge-get-single-lang-server-by-extension filename)
+                                                            (lsp-bridge-get-single-lang-server-by-file-mode filename))))
+                         (if multi-lang-server-by-extension
+                             multi-lang-server-by-extension
+                           lang-server-by-extension))))
+             )))
 
 (defun lsp-bridge-call-async (method &rest args)
   "Call Python EPC function METHOD and ARGS asynchronously."
@@ -1068,7 +1072,7 @@ So we build this macro to restore postion after code format."
       (cond ((member this-command-string '("self-insert-command" "org-self-insert-command"))
              (setq this-command 'acm-filter-insert-char)
              (setq last-command 'acm-filter-insert-char))
-            ((member this-command-string '("delete-block-backward" "grammatical-edit-backward-delete"))
+            ((member this-command-string '("delete-block-backward" "grammatical-edit-backward-delete" "evil-delete-backward-char-and-join"))
              (setq this-command 'acm-filter-delete-char)
              (setq last-command 'acm-filter-delete-char))
             ))))
@@ -1117,11 +1121,12 @@ So we build this macro to restore postion after code format."
             (lsp-bridge-code-action-popup-quit))))
 
       ;; Try send inlay hint if window scroll.
-      (redisplay t) ; NOTE: we need call `redisplay' to force `window-start' return RIGHT line number.
-      (let ((window-pos (window-start)))
-        (when (not (equal lsp-bridge-inlay-hint-last-update-pos window-pos))
-          (lsp-bridge-inlay-hint)
-          (setq-local lsp-bridge-inlay-hint-last-update-pos window-pos)))
+      (when lsp-bridge-enable-inlay-hint
+        (redisplay t) ; NOTE: we need call `redisplay' to force `window-start' return RIGHT line number.
+        (let ((window-pos (window-start)))
+          (when (not (equal lsp-bridge-inlay-hint-last-update-pos window-pos))
+            (lsp-bridge-inlay-hint)
+            (setq-local lsp-bridge-inlay-hint-last-update-pos window-pos))))
       )))
 
 (defun lsp-bridge-close-buffer-file ()
@@ -1154,30 +1159,34 @@ So we build this macro to restore postion after code format."
                                             server-name
                                             completion-trigger-characters
                                             server-names)
-  (lsp-bridge--with-file-buffer filename filehost
-                                ;; Save completion items.
-                                (setq-local acm-backend-lsp-cache-candidates nil)
-                                (setq-local acm-backend-lsp-completion-position position)
-                                (setq-local acm-backend-lsp-completion-trigger-characters completion-trigger-characters)
-                                (setq-local acm-backend-lsp-server-names server-names)
-                                (setq-local acm-backend-lsp-fetch-completion-item-ticker nil)
-
-                                (let* ((lsp-items acm-backend-lsp-items)
-                                       (completion-table (make-hash-table :test 'equal)))
-                                  (dolist (item candidates)
-                                    (plist-put item :annotation (capitalize (plist-get item :icon)))
-                                    (puthash (plist-get item :key) item completion-table))
-                                  (puthash server-name completion-table lsp-items)
-                                  (setq-local acm-backend-lsp-items lsp-items))
-                                (lsp-bridge-try-completion)))
+  ;; avoid completion when cursor move in org
+  (when (or (not (eq major-mode 'org-mode))
+         (eq (plist-get (lsp-bridge--point-position (point)) :character)
+            (plist-get position :character)))
+    (lsp-bridge--with-file-buffer filename filehost
+                                  ;; Save completion items.
+                                  (setq-local acm-backend-lsp-cache-candidates nil)
+                                  (setq-local acm-backend-lsp-completion-position position)
+                                  (setq-local acm-backend-lsp-completion-trigger-characters completion-trigger-characters)
+                                  (setq-local acm-backend-lsp-server-names server-names)
+                                  (setq-local acm-backend-lsp-fetch-completion-item-ticker nil)
+                                  
+                                  (let* ((lsp-items acm-backend-lsp-items)
+                                         (completion-table (make-hash-table :test 'equal)))
+                                    (dolist (item candidates)
+                                      (plist-put item :annotation (capitalize (plist-get item :icon)))
+                                      (puthash (plist-get item :key) item completion-table))
+                                    (puthash server-name completion-table lsp-items)
+                                    (setq-local acm-backend-lsp-items lsp-items))
+                                  (lsp-bridge-try-completion))))
 
 (defun lsp-bridge-try-completion ()
   (cond (lsp-bridge-prohibit-completion
          (setq-local lsp-bridge-prohibit-completion nil))
         (t
          ;; Don't popup completion menu when `lsp-bridge-last-change-position' (cursor before send completion request) is not equal current cursor position.
-         (when (equal lsp-bridge-last-change-position
-                      (list (current-buffer) (buffer-chars-modified-tick) (point)))
+         (when (or t (equal lsp-bridge-last-change-position
+                              (list (current-buffer) (buffer-chars-modified-tick) (point))))
            ;; Try popup completion frame.
            (if (cl-every (lambda (pred)
                            (if (functionp pred)
@@ -1388,6 +1397,8 @@ So we build this macro to restore postion after code format."
   (setq lsp-bridge-last-change-position
         (list (current-buffer) (buffer-chars-modified-tick) (point))))
 
+(defvar-local lsp-org-babel-save-current--point nil)
+
 (defun lsp-bridge-monitor-after-change (begin end length)
   ;; Nothing change actual if `begin' and `end' equal `lsp-bridge--before-change-begin-point' and `lsp-bridge--before-change-end-point'
   ;; Then we should not send any request to search backend.
@@ -1400,6 +1411,9 @@ So we build this macro to restore postion after code format."
           ;; Record last command to `lsp-bridge-last-change-command'.
           (setq lsp-bridge-last-change-command (format "%s" this-command))
 
+          (if (member lsp-bridge-last-change-command '("evil-delete-backward-char-and-join"))
+              (setq-local lsp-bridge-prohibit-completion t))
+
           ;; Record last change position to avoid popup outdate completions.
           (setq lsp-bridge-last-change-position (list (current-buffer) (buffer-chars-modified-tick) (point)))
 
@@ -1409,19 +1423,26 @@ So we build this macro to restore postion after code format."
           ;; Sync change for org babel if we enable it
           (lsp-bridge-org-babel-monitor-after-change begin end length)
 
-          ;; Send LSP requests.
-          (when (or (lsp-bridge-call-file-api-p)
-                    (lsp-bridge-is-remote-file))
-
-            ;; Uncomment below code to debug `change_file' protocol.
-            ;; (message (format "change_file: '%s' '%s' '%s' '%s' '%s' '%s'"
-            ;;                  length
-            ;;                  lsp-bridge--before-change-begin-pos
-            ;;                  lsp-bridge--before-change-end-pos
-            ;;                  (lsp-bridge--position)
-            ;;                  change-text
-            ;;                  (buffer-substring-no-properties (line-beginning-position) (point))
-            ;;                  ))
+          ;; Send change_file request to trigger LSP completion.
+          (when-let ((pos (lsp-bridge--position))
+                     (ok (or (lsp-bridge-call-file-api-p)
+                           (lsp-bridge-is-remote-file))))
+            ;; patch for org babel
+            (when (and (member this-command '(+org/return evil-org-open-below))
+                       (not (eq (plist-get lsp-bridge--before-change-begin-pos :line)
+                                (plist-get lsp-bridge--before-change-end-pos :line)))
+                       lsp-org-babel-save-current--point
+                       (> length 1))
+              (setq end lsp-org-babel-save-current--point)
+              (setq begin lsp-org-babel-save-current--point)
+              (setq length 0)
+              (save-excursion
+                (goto-char lsp-org-babel-save-current--point)
+                (forward-line)
+                (setq end (eol)))
+              (setq lsp-bridge--before-change-begin-pos (lsp-bridge--point-position begin))
+              (setq lsp-bridge--before-change-end-pos (lsp-bridge--point-position begin))
+              (setq pos (lsp-bridge--point-position end)))
 
             ;; Send change_file request to trigger LSP completion.
             (lsp-bridge-call-file-api "change_file"
@@ -1429,7 +1450,7 @@ So we build this macro to restore postion after code format."
                                       lsp-bridge--before-change-end-pos
                                       length
                                       change-text
-                                      (lsp-bridge--position)
+                                      pos
                                       (acm-char-before)
                                       (buffer-name)
                                       (acm-get-input-prefix))
diff --git a/lsp_bridge.py b/lsp_bridge.py
index e2dd85a..cf215cb 100755
--- a/lsp_bridge.py
+++ b/lsp_bridge.py
@@ -766,6 +766,9 @@ class LspBridge:
             if is_remote_path(filepath):
                 return
             open_file_success = True
+            if filepath == '/Users/royokong/.org/roam/.org-src-babel':
+                print('skip org src babel')
+                return
 
             if not is_in_path_dict(FILE_ACTION_DICT, filepath):
                 open_file_success = self.open_file(filepath)  # _do is called inside event_loop, so we can block here.
@@ -867,12 +870,15 @@ def read_lang_server_info(lang_server_path):
 
 def load_single_server_info(lang_server):
     lang_server_info_path = ""
-    if os.path.exists(lang_server) and os.path.dirname(lang_server) != "":
-        # If lang_server is real file path, we load the LSP server configuration from the user specified file.
-        lang_server_info_path = lang_server
-    else:
-        # Otherwise, we load LSP server configuration from file lsp-bridge/langserver/lang_server.json.
-        lang_server_info_path = get_lang_server_path(lang_server)
+    try:
+        if os.path.exists(lang_server) and os.path.dirname(lang_server) != "":
+            # If lang_server is real file path, we load the LSP server configuration from the user specified file.
+            lang_server_info_path = lang_server
+        else:
+            # Otherwise, we load LSP server configuration from file lsp-bridge/langserver/lang_server.json.
+            lang_server_info_path = get_lang_server_path(lang_server)
+    except TypeError:
+        print(f'{lang_server} is error in load_single_server_info')
 
     with open(lang_server_info_path, encoding="utf-8", errors="ignore") as f:
         return read_lang_server_info(f)
diff --git a/multiserver/pyright-background-analysis_ruff.json b/multiserver/pyright-background-analysis_ruff.json
index c5442f3..788c4fa 100644
--- a/multiserver/pyright-background-analysis_ruff.json
+++ b/multiserver/pyright-background-analysis_ruff.json
@@ -1,8 +1,8 @@
 {
   "completion": ["pyright-background-analysis"],
   "completion_item_resolve": ["pyright-background-analysis"],
-  "diagnostics": ["ruff"],
-  "code_action": ["ruff"],
+  "diagnostics": ["pyright", "ruff"],
+  "code_action": ["pyright", "ruff"],
   "execute_command": ["pyright-background-analysis", "ruff"],
   "find_define": "pyright-background-analysis",
   "find_type_define": "pyright-background-analysis",
